arrows/core/generate_texture.h

```diff
@@ -0,0 +1,427 @@
+        if (std::abs(interpolated_depth -  bilinear_interp_safe<double>(depth_maps[i], pt_img(0), pt_img(1))) > depth_threshold)
+        {
+          point_scores[i] = 0;
+        }
+      }
+
+      // adjust the scores for fusion
+      std::fill(values.begin(), values.end(), 0.0);
+      adjust_images_contributions(point_scores);
+      for (size_t i = 0; i < images.size(); ++i)
+      {
+        if (point_scores[i] > 0)
+        {
+          for (size_t d = 0; d < texture.depth(); ++d)
+            values[d] += point_scores[i] * bilinear_interp_safe<T>(images[i], points_2d[i](0), points_2d[i](1), d);
+        }
+      }
+      // fill the texture pixel
+      for (size_t d = 0; d < texture.depth(); ++d)
+        texture.at<T>(x, y, d) = values[d];
+    }
+  }
+}
+
+
+/// This functions dilates the texture atlas using a binary mask
+/**
+ * \param texture [in/out] image to dilate
+ * \param mask [in/out] binary mask used for dilation
+ * \param nb_iter [in] the dilation is repeated nb_iter times
+ */
+template <class T>
+void dilate_atlas(vital::image& texture, vital::image_of<char>& mask, int nb_iter)
+{
+  int height = static_cast<int>(texture.height());
+  int width = static_cast<int>(texture.width());
+
+  // lambda to copy a pixel with any depth
+  auto copy_pixel = [&texture](unsigned int x_d, unsigned int y_d, unsigned int x_s, unsigned int y_s)
+  {
+    T* dest = &texture.at<T>(x_d, y_d, 0);
+    T* src = &texture.at<T>(x_s, y_s, 0);
+    for (int i = 0; i < texture.depth(); ++i, src += texture.d_step(), dest += texture.d_step())
+    {
+      *dest = *src;
+    }
+  };
+  for (int n = 0; n < nb_iter; ++n)
+  {
+    // horizontal dilate
+    for (unsigned int y = 0; y < height; ++y)
+    {
+      for (unsigned int x = 0; x < width - 1; ++x)
+      {
+        char d = mask(x + 1, y) - mask(x, y);
+        if (d == 1)
+        {
+          mask(x, y) = 1;
+          copy_pixel(x, y, x + 1, y);
+        }
+        else if (d == -1)
+        {
+          mask(x + 1, y) = 1;
+          copy_pixel(x + 1, y, x, y);
+          ++x;
+        }
+      }
+    }
+    // vertical dilate
+    for (unsigned int x = 0; x < width; ++x)
+    {
+      for (unsigned int y = 0; y < height - 1; ++y)
+      {
+        char d = mask(x, y + 1) - mask(x, y);
+        if (d == 1)
+        {
+          mask(x, y) = 1;
+          copy_pixel(x, y, x, y + 1);
+        }
+        else if (d == -1)
+        {
+          mask(x, y + 1) = 1;
+          copy_pixel(x, y + 1, x, y);
+          ++y;
+        }
+      }
+    }
+  }
+}
+
+
+/// This function generates a texture from a set of images and maps it on the mesh
+/**
+ * \param mesh [in/out] the mesh to texture.
+ * \param cameras [in] a list of cameras perspective
+ * \param images [in] a list of images
+ * \param resolution [in] resolution of the texture (mesh unit/pixel)
+ */
+template <class T, int N>
```

Comment:
`N` should not be a template parameter here.  `N` is the number of color channels, which we can either get from the input images or pass in as a variables.  Either way, it doesn't need to be fixed at compile time and will just lead to code bloat.